### PR TITLE
serpent: precommit check parameter violation for swiftc has been detected.

### PR DESCRIPTION
**Pull request by lolgear**
### Pull Request Description

- Prevent overriding of custom swiftc path by default `xcrun` output for Darwin.

### Acceptance Criteria

To be accepted into the Swift source compatibility test suite, a project must:

- [x] be an *Xcode* or *swift package manager* project
- [x] support building on either Linux or macOS
- [x] target Linux, macOS, or iOS/tvOS/watchOS device
- [x] be contained in a publicly accessible git repository
- [x] maintain a project branch that builds against Swift 4.0 and passes any unit tests
- [x] have maintainers who will commit to resolve issues in a timely manner
- [x] be compatible with the latest GM/Beta versions of *Xcode* and *swiftpm*
- [x] add value not already included in the suite
- [x] be licensed with one of the following permissive licenses:
	- [x] MIT
- [x] pass `./project_precommit_check` script run

Ensure project meets all listed requirements before submitting a pull request.

### CODE REVIEW

**Comment by lolgear** [comment]
@swift-ci test

**Comment by lolgear** [comment]
@clackary Can you check it?

**Comment by clackary** [comment]
Thanks for the feedback @lolgear! You're right about the parameter being overwritten within the function (at least for Darwin platforms). The swiftc flag is only necessary on non-Darwin platforms where we can't query xcrun for the currently selected swiftc.

I think it should be more clear that this is desired behavior. I've opened #365, which should address this.

**Comment by lolgear** [comment]
@clackary 
Well, I have Xcode-beta in `~/Downloads` folder _which_ I would like to use to check latest Swift, of course.

**Comment by clackary** [comment]
@lolgear 
There are a few ways to do this. `project_precommit_check` is basically just some extra checks around `runner.py`.
If you'd like to use `project_precommit_check`, you'd need to switch to the Xcode in your Downloads folder. Something like this should work... `sudo xcode-select --switch ~/Downloads/Xcode-beta.app`

Or, you could use `runner.py` directly. There are more flags and options here, and should provide some extra flexibility.
```
./runner.py --swift-branch swift-5.1-branch --swiftc ~/Downloads/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc --projects projects.json --include-repos 'path == "Base64CoderSwiftUI"' --include-actions 'action.startswith("Build")'
```

**Comment by lolgear** [comment]
@clackary 
I still don't see any reason to opt-out this parameter on Darwin platforms.
Why I can't just pass this parameter. It will be valid, I suppose, no?
Or give me an opportunity to pass path to `Xcode.app` as a parameter to make your black magic in `./runner.py`

**Comment by clackary** [comment]
I think the original idea was to use `project_precommit_check` to establish the configuration works with current version before testing the master/release branch compilers, which is why it enforced a certain version. If using `runner.py` directly is too difficult, I guess it would make sense to allow `project_precommit_check` to do this.

I've updated the other PR to relax the version check, and provide the functionality you're requesting.

File: project_precommit_check [hunk 1]
```diff
@@ -177,7 +177,13 @@ def is_correct_swift_version(swiftc, compatibility_version):
...
 
 def check(project, swift_branch, swiftc, compatibility_version):
-    if platform.system() == 'Darwin':
+    # Please, rewrite this script.
+    # Writing into function parameters is a bad manner. You can do better.
+    # Please, rewrite it.
+    # Thanks!
+
+    # find pre-installed if we don't specify parameter --swiftc
+    if platform.system() == 'Darwin' and not swiftc:
```
Comment (clackary) [inline project_precommit_check]: This is not sufficient to support `--swiftc` flag on Darwin. The check in `is_correct_swift_version` will fail when trying to use a swiftc that doesn't match the compatibility configuration.

File: project_precommit_check [hunk 2]
```diff
@@ -177,7 +177,13 @@ def is_correct_swift_version(swiftc, compatibility_version):
 
 
 def check(project, swift_branch, swiftc, compatibility_version):
-    if platform.system() == 'Darwin':
+    # Please, rewrite this script.
+    # Writing into function parameters is a bad manner. You can do better.
+    # Please, rewrite it.
+    # Thanks!
```
Comment (clackary) [inline project_precommit_check]: See #365.

Comment (lolgear) [reply project_precommit_check]: @clackary 
I would like to see a message before compatibility version check.

```
    swiftc_path = os.path.abspath(swiftc)
    common.debug_print("checking swiftc_path -> " + swiftc_patch + " and " + comparability_version + " <- compatibility version")
    if not is_correct_swift_version(swiftc_path, compatibility_version):
```

Maybe it is better to add this output to `is_correct_swift_version()`

Comment (clackary) [reply project_precommit_check]: The full path to swifc is already printed in `is_correct_swift_version()`.

https://github.com/apple/swift-source-compat-suite/blob/273551372c4aa5c406b7833991ae77ee39d11bcc/project_precommit_check#L160

**Comment by clackary** [comment]
@lolgear #365 has merged, which should give you the behavior you requested. Thanks for the feedback here!

**Pull request closed without merging**